I like rust (#1)
thomas-zahner/website: content/posts/i-like-rust.md

+++
title = "I like Rust"
date = 2024-10-02

[taxonomies]
tags = ["rust"]
+++

If you were to ask me what my favourite programming language is, my answer would be Rust. For the majority of projects I'd prefer Rust to any other language I know. This is not because I simply like the syntax of Rust, but because with Rust I can write code quickly and with confidence that it will work. More so than with any other language.

<!-- more -->

## Rust is correct

First of all, Rust as a language deeply cares about correctness and doing things the right way.
This also applies to the ecosystem surrounding the language such as the standard library and the crates on crates.io.

The roots for this sense of correctness might lie in safety guarantees that Rust provides as a memory safe language. Violating ownership, type or lifetime constraints would potentially lead to undefined behaviour so that Rust has to be inherently pedantic about enforcing such rules in order to be memory safe. But when you work with the Rust ecosystem, you realise that correctness doesn't end with memory safety. It runs throughout Rust's narrative.

> Rust inherited many ideas from Haskell and declarative languages in general. Declarative languages come from academia and mathematics. They are what you get when you combine mathematics and logic with computer science and programming languages. For example, they allow to prove the correctness of a program without running it. In functional programming certain things, like modifying lists, can be done much more elegantly with less code and no bugs. The disadvantage of functional programming is that interaction with the real world, such as talking to hardware and reading user input, is cumbersome. Rust has somehow managed to be great at both things. Rust programs can and are formally verified in many ways. The compiler verifies types, ownership, lifetimes and more. You can additionally verify that your program [never panics](https://github.com/dtolnay/no-panic) or that it [doesn't encounter undefined behaviour](https://github.com/rust-lang/miri).

Reviewer: Comma missing:

"In functional programming, certain things, like modifying lists, can be done much more elegantly with less code and no bugs."

Reviewer: "Rust programs can and are formally verified in many ways."
This is partially true, but it might be overstating the case. Formal verification is possible for Rust programs, but it's not a built-in feature of the language or commonly used in most Rust development. There are tools and projects working on formal verification for Rust, but it's not as widespread or integrated as this sentence might imply.
Maybe just "Rust's compiler does perform extensive static analysis"?

> Rustaceans seem to obsess over seemingly simple things like semantic versioning (SemVer). Developers of Rust libraries won't release the first major version (1.\*.\*) unless they are sure that their API is mature and stable enough. This process normally takes years and has also kind of become a meme. Correct SemVer also means no incompatible API changes with the release of patch (\*.\*.1) and minor (\*.1.\*) versions. Complying with those seemingly simple rules turns out to be pretty hard in practice. This is why [linters are created](https://github.com/obi1kenobi/cargo-semver-checks) to prevent authors from making mistakes.

Reviewer: > Correct SemVer also means no incompatible API changes with the release of patch (\*.\*.1) and minor (\*.1.\*) versions.

Yes, unless we're talking about pre-1.0 releases. 😉 It's a minor but important differentiation.

> When you work with functions and types from the standard library, you can feel just how much thought and polish has gone into the design. This vastly contrasts with most other languages I've used such as JavaScript or Java.

Reviewer: I'd move this part up; perhaps after "It runs throughout Rust's narrative."
The reason being that it would make for a nice transition from Rust's design philosophy over the standard library all the way to third-party libraries. It permeates the entire ecosystem.

## An example

> Recently, I had to deal with dates and date-times in Java. I wanted to validate a string representing a date or date with time and have a type to encapsulate the information. The first type you encounter when looking this will be `java.util.Date`. You will notice quickly that most methods on this type are marked as deprecated. When reading about it you learn how it doesn't represent a date but an instance in time measured in milliseconds. Basically the naming was bad and the implementation of this type was inconsistent, poorly designed and thread-unsafe.[^1] This date type was introduced in version 1.0 (1996) and most methods were deprecated in 1.1 (2002).[^2] So today this type only exists for backwards compatibility and to confuse new developers. Instantiating this type can for example be done by providing year, month and day. But remember that the year starts at 1900, the first month is 0 and the first day is 1. Month and day overflows also seem to be a handy feature.

Reviewer: "when looking this" -> "when looking into this"

> Recently, I had to deal with dates and date-times in Java. I wanted to validate a string representing a date or date with time and have a type to encapsulate the information. The first type you encounter when looking this will be `java.util.Date`. You will notice quickly that most methods on this type are marked as deprecated. When reading about it you learn how it doesn't represent a date but an instance in time measured in milliseconds. Basically the naming was bad and the implementation of this type was inconsistent, poorly designed and thread-unsafe.[^1] This date type was introduced in version 1.0 (1996) and most methods were deprecated in 1.1 (2002).[^2] So today this type only exists for backwards compatibility and to confuse new developers. Instantiating this type can for example be done by providing year, month and day. But remember that the year starts at 1900, the first month is 0 and the first day is 1. Month and day overflows also seem to be a handy feature.

Reviewer: "the naming was bad" -> "the naming is misleading"

> Recently, I had to deal with dates and date-times in Java. I wanted to validate a string representing a date or date with time and have a type to encapsulate the information. The first type you encounter when looking this will be `java.util.Date`. You will notice quickly that most methods on this type are marked as deprecated. When reading about it you learn how it doesn't represent a date but an instance in time measured in milliseconds. Basically the naming was bad and the implementation of this type was inconsistent, poorly designed and thread-unsafe.[^1] This date type was introduced in version 1.0 (1996) and most methods were deprecated in 1.1 (2002).[^2] So today this type only exists for backwards compatibility and to confuse new developers. Instantiating this type can for example be done by providing year, month and day. But remember that the year starts at 1900, the first month is 0 and the first day is 1. Month and day overflows also seem to be a handy feature.

Reviewer: > and to confuse new developers

I find that a bit snarky. Maybe just "and it's a constant source of confusion"? Or, if you want to keep it but add a humorous touch, perhaps consider "and it's haunting new developers to this day."

> Recently, I had to deal with dates and date-times in Java. I wanted to validate a string representing a date or date with time and have a type to encapsulate the information. The first type you encounter when looking this will be `java.util.Date`. You will notice quickly that most methods on this type are marked as deprecated. When reading about it you learn how it doesn't represent a date but an instance in time measured in milliseconds. Basically the naming was bad and the implementation of this type was inconsistent, poorly designed and thread-unsafe.[^1] This date type was introduced in version 1.0 (1996) and most methods were deprecated in 1.1 (2002).[^2] So today this type only exists for backwards compatibility and to confuse new developers. Instantiating this type can for example be done by providing year, month and day. But remember that the year starts at 1900, the first month is 0 and the first day is 1. Month and day overflows also seem to be a handy feature.

Reviewer: > Instantiating this type can for example be done by providing year, month and day.

Instantiating this type can, for example, be done by providing year, month and day.

> Recently, I had to deal with dates and date-times in Java. I wanted to validate a string representing a date or date with time and have a type to encapsulate the information. The first type you encounter when looking this will be `java.util.Date`. You will notice quickly that most methods on this type are marked as deprecated. When reading about it you learn how it doesn't represent a date but an instance in time measured in milliseconds. Basically the naming was bad and the implementation of this type was inconsistent, poorly designed and thread-unsafe.[^1] This date type was introduced in version 1.0 (1996) and most methods were deprecated in 1.1 (2002).[^2] So today this type only exists for backwards compatibility and to confuse new developers. Instantiating this type can for example be done by providing year, month and day. But remember that the year starts at 1900, the first month is 0 and the first day is 1. Month and day overflows also seem to be a handy feature.

Reviewer: > But remember that the year starts at 1900, the first month is 0 and the first day is 1. Month and day overflows also seem to be a handy feature.

wtf. 😆

Author: Haha yeah Java things :rofl:

```java
new Date(0, 0, 0) // Sun Dec 31 00:00:00 CET 1899
new Date(0, 33, 0); // Tue Sep 30 00:00:00 CET 1902 
```

The implementation was terrible. After deprecation, you would use third-party libraries instead. This was until Java 8 (or 1.8, Java versions don't make sense) where it was decided to reimplement the date type into the standard library by copying the now popular and established libraries. This implementation was indeed much better, however I still noticed inaccuracies. For example you won't notice when an invalid date in the following case is provided.

```java
DateTimeFormatter f = DateTimeFormatter.ofPattern("yyyy-MM-dd");

// Official documentation
// Throws:
//    DateTimeParseException - if the text cannot be parsed
LocalDate d = LocalDate.parse("2021-02-31", f);

// ah never mind, the date is silently rounded off in this case... 
System.out.println(d); // 2021-02-28
```

> In stark contrast to Java, the Rust standard library is meticulously correct and polished. Rust developers learned form the mistakes made in the past. Rust is conservative when it comes to adding new features. Only once they have proven useful in other languages or in popular libraries they are considered to be added.

Reviewer: learned form -> learned from

## Technical debt

The previous example shows how the concept of technical debt also applies to programming languages.

> In fact, debts in language design weigh much heavier.

Reviewer: ```suggestion
In fact, debt in language design weighs much heavier.
```

> Once a language has commited a mistake, it cannot be undone without breaking backwards compatibility.

Reviewer: ```suggestion
Once a language has committed a mistake, it cannot be undone without breaking backward compatibility.
```

Some established languages have evolved significantly over time.

> There have been many additions when new conepts and approaches got popular.

Reviewer: This sentence is confusing me. I'm also not sure if it's necessary.

> Some new approaches replaced existing ones, but they couldn't be removed because of the need to maintain backwards compatibility.

Reviewer: ```suggestion
New approaches were introduced to replace existing ones, but the old approaches couldn't be removed due to the need to maintain backward compatibility.
```

This sentence was confusing me, too. I didn't know what "it" referred to in the second part of the sentence.

This leads to conventions and best practices.
Previous best practices become outdated and frowned upon.
The language accumulates complexity as its technical debt continues to grows.

C++, for example, has become so complicated over time that "modern C++" conventions have been created
where only a subset of the language should be used.
Although it is natural for a language to evolve to some extent, too much change can be unhealthy.
My feeling about Rust is that its foundations are so robust that it won't change as radically as other languages have over time.

## Types

> Personally, I think it is a bad decision to use dynamically typed languages for anything other than small projects or scripting. Look at any popular dynamically typed language ([Python](https://docs.python.org/3/library/typing.html), [PHP](https://www.php.net/manual/en/language.types.declarations.php), [JavaScript](https://www.typescriptlang.org/), ...) and you will see the same trend; the desire for more type safety. Unfortunately, they will never be able to achieve type safety.

Reviewer: ```suggestion
Personally, I think it is a bad decision to use dynamically typed languages for anything other than small projects or scripting. Look at any popular dynamically typed language ([Python](https://docs.python.org/3/library/typing.html), [PHP](https://www.php.net/manual/en/language.types.declarations.php), [JavaScript](https://www.javascript.com/), ...) and you will see the same trend; the desire for more type safety. Unfortunately, they will never be able to achieve type safety.
```

I am aware that JS has no official homepage, but the link from JavaScript to TypeScript is a bit misleading. How about linking to this instead?

Author: Ah, here I actually tried to link to the type systems the languages introduced, in JavaScript this is now mainly done with TypeScript. But I'll replace it with footnotes then to make it more clear

It might be obvious that problems caught in production are worse than problems noticed during implementation.
But the effective cost difference is huge.
IBM showed this in [a study](https://www.researchgate.net/publication/255965523_Integrating_Software_Assurance_into_the_Software_Development_Life_Cycle_SDLC) conducted in 2010.

![IBM System Science Institute. A graph depicting the relative cost of fixing defects. "Design" has a cost of 1, "implementation" a cost of 6.5, "testing" a cost of 15 and "maintenance" a cost of 100.](/ibm-relative-cost-of-fixing-defects.png)

> Rust is as statically typed and type safe as you could possibly get with a "real" programming language. Static types don't automatically lead to type safety. For example in Java you can do the following to trick the compiler and cause a runtime error:

Reviewer: What do you mean by "real" here? Mainstream?

Some languages, particularly those based on dependent type theory (like Idris or Coq), can provide even stronger type safety guarantees. So you probably want to be precise with your wording here. 😉

```java
int x = 42;
Object forgetCompileTypeInfo = x; // everything is an object
return (String) forgetCompileTypeInfo; // oh no, a runtime error
```

In Rust that's not possible. Casts must be valid at the type level, or else they will be prevented statically.[^3] Not checking for type constraints at compile time, or worse, doing conversions at runtime to somehow satisfy type constraints, as JavaScript does[^4], leads to more bugs and costs.

> ## No billion dollar bug

Reviewer: Hoare called it the billion dollar mistake, according to Wikipedia: https://en.wikipedia.org/wiki/Tony_Hoare

> But I couldn't resist the temptation to put in a null reference, simply because it was so easy to implement. This has led to innumerable errors, vulnerabilities, and system crashes, which have probably caused a billion dollars of pain and damage in the last forty years.[^5]

Rust is one of the languages which does not know a special null value. You must opt in for the possible absence of a value. This is done with the `Option<T>` type.

## Error handling

> Most programmers today agree that using the GOTO statement in high-level languages is bad practice. It leads code that is more difficult to read and maintain. I think exceptions like they are known in many modern languages are just fancy GOTO statements. When an exception is hit, the normal execution flow is stopped and the instruction pointer jumps into some error handling section.

Reviewer: > Most programmers today agree that using the GOTO statement in high-level languages is bad practice.

Maybe link to the paper? https://homepages.cwi.nl/~storm/teaching/reader/Dijkstra68.pdf

In a sense, Rust only knows "checked" exceptions but with a very different syntax.
Instead of throwing an exception we simply return a `Result<T, E>`.
The result's value can be either `Ok(T)` ("`Ok` with a value of `T`) or `Err(E)`.
You will be forced to check if the result is `Ok` or `Err` if you want to access the inner values `T` or `E`.
Now there are certain scenarios where your program reaches a state where you can't or don't want to recover, when encountering a problem.
In this case it's possible to `panic`. When a program panics it will crash by unwinding its call-stack, optionally printing a stack trace.

> You won't be able to catch panics.

Reviewer: In Rust, you can indeed catch panics.

Globally, with https://doc.rust-lang.org/std/panic/fn.set_hook.html
Locally, with https://doc.rust-lang.org/beta/std/panic/fn.catch_unwind.html

Author: Ah didn't know that, thanks. I'll rephrase it then

So Rust's approach is simple and pragmatic. A fault condition should be either recoverable or not.

## Error messages

> Error messages in Rust are awesome. They're super helpful, concise and correct. As the compiler itself is very strict in general you will encounter compilation errors where you don't know how to resolve the issue and you have to rethink your approach. But the error messages are as helpful as they could possible get. An unspecific or unclear error message is considered a bug by the Rust team, and you should report it if you ever encounter one. Some compilers give vague error messages and some even point to the wrong source files! This for example happens when working with templates in C++. Working with such error messages is a nightmare.

Reviewer: could possible get -> could possibly get

Runtime errors in Rust are a rarer phenomenon but they will be just as clear as the error messages by the compiler. Most of the time the concise messages are enough to track down the mistake. Optionally backtraces can be enabled as Rust provides a tiny runtime.

If you create a stack overflow with C++ the error message could look like this[^6]:

```
HelloWorld.exe (process 15916) exited with code -1073741571.
```

Or it might look like this:

```
Segmentation fault
```

With Rust the error message looks as follows:

```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
```

With Rust you don't have to be a coding wizard to understand error messages. In my attempt to create a stack overflow in Rust, the compiler also warned me about the fact that my `function cannot return without recursing`.

## Dependencies

> When writing software these days, you will probably sooner or later depend on external libraries. This is why dependency management shouldn't be a pain. Many languages evolved before streamlined dependency management was a thing. I think NPM, the package manager for NodeJS was the first good dependency management system for a programming language. I became very popular and might be the main reason why NodeJS itself got so popular. One thing I dislike about the NPM ecosystem is that dependency graphs tend to be huge. This is not a problem of NPM itself. To some extent it might be a problem of NodeJS or JavaScript. Simple things like cloning objects or comparing objects for equality are not possible out of the box. When people in an ecosystem tend to add dependencies to their libraries without much thought it becomes the norm. Rustaceans seem to be more conservative about adding dependencies to their projects. Rust's build system and package manager, Cargo, is very similar to NPM with a few improvements. With NPM, for example, I regularly forget to install or update the dependencies with `npm install`, whereas Cargo does this automatically for you.

Reviewer: > I think NPM, the package manager for NodeJS was the first good dependency management system for a programming language.

Define "good". 😆 
NPM was first released in 2010.
Both, Python's pip (2008) and Ruby gems (2004) predate it.

> When writing software these days, you will probably sooner or later depend on external libraries. This is why dependency management shouldn't be a pain. Many languages evolved before streamlined dependency management was a thing. I think NPM, the package manager for NodeJS was the first good dependency management system for a programming language. I became very popular and might be the main reason why NodeJS itself got so popular. One thing I dislike about the NPM ecosystem is that dependency graphs tend to be huge. This is not a problem of NPM itself. To some extent it might be a problem of NodeJS or JavaScript. Simple things like cloning objects or comparing objects for equality are not possible out of the box. When people in an ecosystem tend to add dependencies to their libraries without much thought it becomes the norm. Rustaceans seem to be more conservative about adding dependencies to their projects. Rust's build system and package manager, Cargo, is very similar to NPM with a few improvements. With NPM, for example, I regularly forget to install or update the dependencies with `npm install`, whereas Cargo does this automatically for you.

Reviewer: > Rustaceans seem to be more conservative about adding dependencies to their projects.

Huh, I've seen a similar attitude towards adding dependencies as in Node...
At least, it's a bold statement, which I'd back up with data or a reference.

You just mark a structure with `Clone` or `Eq`, that's one simple line of code.
You like linters? Rust comes with [Clippy <img src="/clippy.webp" height="25" style="border: none;">](https://doc.rust-lang.org/stable/clippy/usage.html),
the most strict and annoying linter.
Have you made enemies by fighting over code formatting rules?
In Rust, the war is over because there is the [official style guide](https://doc.rust-lang.org/beta/style-guide/index.html),
which you never have to think about because your IDE or editor uses rustfmt anyway.

## Compilation times

When people complain about Rust they often mention compile times.
It's true that compilation of Rust projects do take their time.
But compiling C++ or Java projects really isn't any faster.
Given Rust's abstractions and complex rules, I'm actually impressed that Rust is on par with other languages.
In practice, compile times rarely bother me.
I recompile my Rust programs less, because more bugs are caught by rust-analyser in my IDE, rather than during test execution.
Compile times have improved over the last years and they will continue to improve.

> If you have issues with compilation times there are [many ways to further improve them](https://corrode.dev/blog/tips-for-faster-rust-compile-times/).

Reviewer: 😊

## Hard to learn

The Rust compiler is strict, that's no secret. As a beginner it can be difficult to get basic programs to compile. Understanding the concepts of the language such as ownership is therefore essential. Fortunately, the official [Rust book](https://doc.rust-lang.org/stable/book/) is perfect for learning all these concepts, and the Rust compiler will also help you along the way.

You will encounter compilation errors probably more often than in any other language.
There will be cases where you have no idea how to fix the problem at first.
Eventually, you might realise that you will have to rethink your whole approach to the problem.

This was the hardest part for me (and still is) when working with Rust. Sometimes the compiler forces you to rethink your approach. You realise that you can't bend the borrow checker to your will, instead you have to work with it. The compiler is your mentor not your enemy to fight.

> Personally, I started learning Rust during the Covid pandemic. At the time we had a course on algorithms and data structures at university and I decided to implement the exercises in Rust instead of Java. After reading the first few chapters of the Rust book, I felt confident enough to give it a try. I was able to implement some algorithms and was happy with the results. But I got lost and confused when I tried to implement a binary tree myself. After trail and error, I did some research and came across a chapter in the book I hadn't read yet. Turns out that data structures in Rust can get complicated very quickly. On top of ownership you also need to know about smart pointers, interior mutability, reference counting, reference cycles...

Reviewer: trail and error -> trial and error

I was overwhelmed and discouraged. But I realised that data structures were not the best starting point for learning Rust. And that in the real world you would make use of the standard library and crates built by smart people anyway.

> After that, I tried to implement simple things with Rust. Over time I almost forgot about the concepts I read about when dealing with binary trees. It turns out that programming "normal" things in Rust is much easier than I though. The only thing you really need to understand is ownership. Understanding lifetimes, interior mutability, etc. is great, but in "normal" code you almost never make use of these concepts.

Reviewer: than I though -> than I thought

## Conclusion

In my opinion, Rust is the pinnacle of modern compiler and language development.
The deliberate design choices in the language show how the developers have learned from the problems in other languages.

> The introduction of new concepts such as the borrow checker make the Rust compiler probably the most complex compiler out there.

Reviewer: Hm, maybe? Then again, the C++ compiler is probably also pretty complex. Also, `rustc` is just a compiler frontend for LLVM, so that would be worth a mention as well. ¯\_(ツ)_/¯
